README.md
# autosign
[![Build Status](https://travis-ci.org/danieldreier/autosign.svg?branch=master)](https://travis-ci.org/danieldreier/autosign) [![Code Climate](https://codeclimate.com/github/danieldreier/autosign/badges/gpa.svg)](https://codeclimate.com/github/danieldreier/autosign) [![Dependency Status](https://gemnasium.com/danieldreier/autosign.svg)](https://gemnasium.com/danieldreier/autosign) [![Yard Docs](http://img.shields.io/badge/yard-docs-blue.svg)](http://rubydoc.info/github/danieldreier/autosign) [![Inline docs](http://inch-ci.org/github/danieldreier/autosign.png)](http://inch-ci.org/github/danieldreier/autosign) [![Coverage Status](https://coveralls.io/repos/danieldreier/autosign/badge.svg?branch=master&service=github)](https://coveralls.io/github/danieldreier/autosign?branch=master) [![Gem Version](https://badge.fury.io/rb/autosign.svg)](http://badge.fury.io/rb/autosign)

Tooling to make puppet autosigning easy, secure, and extensible

### Introduction

This tool provides a CLI for performing puppet policy-based autosigning using JWT tokens. Read more at https://danieldreier.github.io/autosign.

### Quick Start: How to Generate Tokens

##### 1. Install Gem on Puppet Master
```shell
gem install autosign
```

##### 2. Generate default configuration

```shell
autosign config setup
```

##### 3. Generate your first autosign token on the puppet master
```shell
autosign generate foo.example.com
```

The output will look something like
```
Autosign token for: foo.example.com, valid until: 2015-07-16 16:25:50 -0700
To use the token, put the following in ${puppet_confdir}/csr_attributes.yaml prior to running puppet agent for the first time:

custom_attributes:
  challengePassword: "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzUxMiJ9.eyJkYXRhIjoie1wiY2VydG5hbWVcIjpcImZvby5leGFtcGxlLmNvbVwiLFwicmVxdWVzdGVyXCI6XCJEYW5pZWxzLU1hY0Jvb2stUHJvLTIubG9jYWxcIixcInJldXNhYmxlXCI6ZmFsc2UsXCJ2YWxpZGZvclwiOjcyMDAsXCJ1dWlkXCI6XCJkM2YyNzI0OC1jZDFmLTRhZmItYjI0MC02ZjBjMDU4NWJiZDNcIn0iLCJleHAiOiIxNDM3MDg5MTUwIn0.lC-EzWaV2dL81aLL7P-9mGwNbiOQDJWcoYjuSHVOqmaLtc7Wis5OZvHFOLln2Fn9qv98oSTnZsIkjmFpbI5dvA"
  ```

The resulting output can be copied to `/etc/puppet/csr_attributes.yaml` on an agent machine prior to running puppet for the first time to add the token to the CSR as the `challengePassword` OID. (just copy-paste from one terminal to another to copy the text)

### Quick Start: Puppet Master Configuration

Run through the previous quick start steps to get the gem installed, then configure puppet to use the `autosign-validator` executable as the policy autosign command:

##### 1. Prerequisities
Note that these settings will be slightly different if you're running Puppet Enterprise, because you'll need to use the `pe-puppet` user instead of `puppet`.

```shell
mkdir /var/autosign
chown puppet:puppet /var/autosign
chmod 750 /var/autosign
touch /var/log/autosign.log
chown puppet:puppet /var/log/autosign.log
```


##### 2. Configure master
```shell
puppet config set autosign $(which autosign-validator) --section master
```

Your master is now configured to autosign using the autosign gem. 


##### 3. Using Legacy Autosign Scripts

If you already had an autosign script you want to continue using, add a setting to your `autosign.conf` like:

```yaml
multiplexer:
  external_policy_executable: "/path/to/autosign/executable"
```

The master will validate the certificate if either the token validator or the external validator succeeds.

If the autosign script was just validating simple strings, you can use the `password_list` validator instead. For example, to configure the master to sign any CSR that includes the challenge passwords of "hunter2" or "CPE1704TKS" you would add:

```ini
password_list:
  password: "hunter2"
  password: "CPE1704TKS"
```

Note that this is a relatively insecure way to do certificate autosigning. Using one-time tokens via the `autosign generate` command is more secure. This functionality is provided to grandfather in existing use cases to ease the transition.

## Validation order
By default the validation runs the following validators in order: 

1. jwt_token
2. password_list
3. multiplexer

The first validator to succeed wins and short circuits the validaiton process.

You can completely customize the list and how they are ordered via the configuration file.  Or even remove some entirely.

```
---
general:
  loglevel: debug
  logfile: "/var/log/autosign.log"
  validation_order: 
    - jwt_token
    - multiplexer
    - password_list
jwt_token:
  secret: J7/WjmkC/CJp2K0/8+sktzSgCqQ=
  validity: '7200'
  journalfile: "/root/var/autosign/autosign.journal"
```

The validation_order config is an ordered array and since the validators will only match the first validation
to succeed the validation script should occur as fast as you want.  

Additionally, if you omit any validator that validator will not be used during the validation process.  This might 
be important if you wanted to only use special validators or remove unwanted validator execution.

Please note, the name of the validator which is speficed by the `NAME` constant in the validator code must match
the list you specify otherwise it will not be part of the validation process.

**NOTE** To use this feature you must have deep_merge 1.2.1+ installed which is now a requirement of this gem.

### Troubleshooting
If you're having problems, try the following:

- Set `loglevel: "debug"` in `/etc/autosign.conf`
- Check the `journalfile`, in `/var/autosign/autosign.journal` by default, to see if the one-time token's UUID has already been recorded. It's just YAML, so you can either delete it or remove the offending entry if you actually want to re-use a token.
- you can manually trigger the autosigning script with something like `cat the_csr.csr | autosign-validator certname.example.com`
- If you run the puppet master foregrounded, you'll see quite a bit of autosign script output if autosign loglevel is set to debug.

Starting with the 1.0.0 release the autosign gem requires ruby 2.4.  If you can't upgrade just yet you can continue to use the older 0.1.4 release.

### Further Reading

- [https://danieldreier.github.io/autosign](https://danieldreier.github.io/autosign) has background on why this exists.
- Automatically generated code documentation in YARDOC format is [available on rubydoc.info](http://rubydoc.info/github/danieldreier/autosign).
- Look at the [puppet-autosign](https://travis-ci.org/danieldreier/puppet-autosign) puppet module to automate setup of this tool, and for a puppet function to generate tokens inside of Puppet, for example when provisioning systems in AWS.


CHANGELOG.md
# Augosign changelog

## Unreleased

## 1.0.1
* (maint) Remove gem dependency on 2.4

## 1.0.0
Released May 19, 2020 

* (maint) print config in yaml format
* (maint) Fix a cache bug where the settings were loaded multiple times
* (maint) Fix a bug where the config overwrites settings
* (maint) Fix bug with validation order
* (maint) The autosign gem now requires the deep_merge gem 1.2.1
* (maint) The require_all has been dropped and is no longer a dependency
* (maint) Fix deprecation warnings with gemspec file
* (maint) Objectify the validator classes
* (maint) Fix a cache bug where the settings were loaded multiple times.
    This was causing overzealous logging
* (feat) Add an ordered validator list
* (feat) Any validator should short circuit
* (feat) Allow user to specify validation order

This release removes support for ruby < 2.4.

## 0.1.4
Released Nov 25, 2019 

### Bug fixes

* Use multi_json to allow a variety of JSON engines to be used, which makes installation easier.
* Read all of STDIN regardless of whether we’ll use it in order to avoid a bug in Java 8.
* Change yard from a runtime dependency to a dev dependency.
* Security updates for dependencies:
* Bump ffi from 1.9.10 to 1.9.25
* Bump yard from 0.9.12 to 0.9.20

## 0.1.3
Released Jan 24, 2018

### Bug fixes

* Fix config file path; the latest version of puppet-autosign creates config files in /etc/puppetlabs/puppetserver/autosign.conf but we weren't checking there
* @reidmv fixed a bug where the decoder would error when presented with a csr with no challengePassword
* added an Apache license to be explicit about how the code is licensed. Did check with all contributors first.

## 0.1.1
Released Oct 30, 2015

* bump version to 0.1.1 to fix safe_yaml issue

## 0.0.6 
Released Jul 15, 2015

* add autosign-validator executable to gem


lib/autosign/config.rb
# frozen_string_literal: true

require 'rbconfig'
require 'securerandom'
require 'deep_merge'
require 'yaml'

module Autosign
  # Exceptions namespace for Autosign class
  module Exceptions
    # Exception representing a general failure during validation
    class Validation < RuntimeError
    end
    # Exception representing a missing file during config validation
    class NotFound < RuntimeError
    end
    # Exception representing a permissions error during config validation
    class Permissions < RuntimeError
    end
    # Exception representing errors that Autosign does not know how to handle
    class Error < RuntimeError
    end
  end

  # Class to manage configuration settings.
  # The purpose of this class is to interact with the configuration file,
  # merge defaults and user-provided settings, and present a configuration hash
  # so that validators and other components do not have to re-implement config
  # file handling.
  class Config
    # Create a config instance to interact with configuration settings
    # To specify a configuration file, settings_param should include something like:
    # {'config_file' => '/usr/local/etc/autosign.conf'}
    #
    # If no defaults are provided, the class checks several common locations for config file path.
    #
    # @param settings_param [Hash] config settings that should override defaults and config file settings
    # @return [Autosign::Config] instance of the Autosign::Config class
    def initialize(settings_param = {})
      # set up logging
      @log = Logging.logger[self.class]
      @log.debug "initializing #{self.class.name}"
      # validate parameter
      raise 'settings is not a hash' unless settings_param.is_a?(Hash)

      # look in the following places for a config file
      @config_file_paths = ['/etc/puppetlabs/puppetserver/autosign.conf', '/etc/autosign.conf', '/usr/local/etc/autosign.conf']

      # HOME is unset when puppet runs, so we need to only use it if it's set
      unless ENV['HOME'].nil?
        @config_file_paths << File.join(Dir.home, '.autosign.conf')
      end

      unless settings_param['config_file'].nil?
        @config_file_paths = [settings_param['config_file']]
      end

      @settings = settings_param
      @log.debug 'Using merged settings hash: ' + @settings.to_s
    end

    # Return a merged settings hash of defaults, config file settings
    # and passed in settings (such as from the CLI)
    #
    # @return [Hash] deep merged settings hash
    def settings
      @log.debug 'merging settings'
      setting_sources = [default_settings, configfile, @settings]
      merged_settings = setting_sources.inject({}) { |merged, hash| merged.deep_merge!(hash, {:overwrite_arrays => true}) }
      @log.debug 'using merged settings: ' + merged_settings.to_s
      merged_settings
    end

    private

    # default settings hash for the whole autosign gem
    # the format must map to something an ini file can represent, so the
    # structure should have a top-level key named "general" and key-value pairs
    # below it, with strings as keys and strings or integers as values.
    #
    # validator settings should not be here; put those in the validator's
    # default settings method.
    #
    # @return [Hash] default configuration settings
    def default_settings
      { 'general' =>
        {
          'loglevel' => 'INFO',
          'validation_order' => %w[
            jwt_token password_list multiplexer
          ]
        },
        'jwt_token' => {
          'validity' => 7200
        } }
    end

    # Locate the configuration file, parse it from INI-format, and return
    # the results as a hash. Returns an empty hash if no config file is found.
    #
    # @return [Hash] configuration settings loaded from INI file
    def configfile
      @log.debug 'Finding config file'
      @config_file_paths.each do |file|
        @log.debug "Checking if file '#{file}' exists"
        if File.file?(file)
          @log.debug 'Reading config file from: ' + file
          config_file = File.read(file)
          parsed_config_file = YAML.load(config_file)
          # parsed_config_file = IniParse.parse(config_file).to_hash
          @log.debug 'configuration read from config file: ' + parsed_config_file.to_s
          return parsed_config_file if parsed_config_file.is_a?(Hash)
        else
          @log.debug "Configuration file '#{file}' not found"
        end
      end
      {}
    end

    # Validate configuration file
    # Raises an exception if the config file cannot be validated
    #
    # @param configfile [String] the absolute path of the config file to validate
    # @return [String] the absolute path of the config file
    def validate_config_file(configfile = location)
      @log.debug 'validating config file'
      unless File.file?(configfile)
        @log.error "configuration file not found at: #{configfile}"
        raise Autosign::Exceptions::NotFound
      end

      # check if file is world-readable
      if File.world_readable?(configfile) || File.world_writable?(configfile)
        @log.error "configuration file #{configfile} is world-readable or world-writable, which is a security risk"
        raise Autosign::Exceptions::Permissions
      end

      configfile
    end

    # Generate a default configuration file
    # As a convenience for the user, we can generate a default config file
    # This class is currently too tightly coupled with the JWT token validator
    def self.generate_default(settings_param = {})
      os_defaults = (
        case RbConfig::CONFIG['host_os']
        when /darwin|mac os/
          {
            'logpath' => File.join(Dir.home, 'autosign.log'),
            'confpath' => File.join(Dir.home, '.autosign.conf'),
            'journalfile' => File.join(Dir.home, '.autosign.journal')
          }
        when /linux/
          {
            'logpath' => '/var/log/autosign.log',
            'confpath' => '/etc/autosign.conf',
            'journalfile' => File.join(Dir.home, '/var/autosign/autosign.journal')
          }
        when /bsd/
          {
            'logpath' => '/var/log/autosign.log',
            'confpath' => '/usr/local/etc/autosign.conf',
            'journalfile' => File.join(Dir.home, '/var/autosign/autosign.journal')
          }
        else
          raise Autosign::Exceptions::Error, "unsupported os: #{host_os.inspect}"
        end
      )

      config = {
        'general' => {
          'loglevel' => 'warn',
          'logfile' => os_defaults['logpath'],
          'validation_order' => %w[
            jwt_token password_list multiplexer
          ]
        },
        'jwt_token' => {
          'secret' => SecureRandom.base64(20),
          'validity' => '7200',
          'journalfile' => os_defaults['journalfile']
        }
      }

      #      config = IniParse.gen do |doc|
      #        doc.section("general") do |general|
      #          general.option("loglevel", "warn")
      #          general.option("logfile", os_defaults['logpath'])
      #        end
      #        doc.section("jwt_token") do |jwt_token|
      #          jwt_token.option("secret", SecureRandom.base64(15))
      #          jwt_token.option("validity", 7200)
      #          jwt_token.option("journalfile", os_defaults['journalfile'])
      #        end
      #        doc.section("multiplexer") do |jwt_token|
      #          jwt_token.option(";external_policy_executable", '/usr/local/bin/some_autosign_executable')
      #          jwt_token.option(";external_policy_executable", '/usr/local/bin/another_autosign_executable')
      #        end
      #        doc.section("password_list") do |jwt_token|
      #          jwt_token.option(";password", 'static_autosign_password_here')
      #          jwt_token.option(";password", 'another_static_autosign_password')
      #        end
      #      end.to_ini
      config_file = settings_param['config_file'] || os_defaults['confpath']
      if File.file?(config_file)
        raise Autosign::Exceptions::Error, "file #{config_file} already exists, aborting"
      end
      return config_file if File.write(config_file, config.to_yaml)
    end
  end
end
